File: packages/allstak-python/src/allstak/integrations/auto_breadcrumbs.py
"""Automatic breadcrumb instrumentation for common Python libraries."""

from __future__ import annotations

import functools
import logging
from typing import Any, Callable, Dict, Optional

logger = logging.getLogger("allstak.sdk")

_SENSITIVE_PARAMS = {"token", "key", "secret", "password", "auth", "api_key"}


def _sanitize_url(url: str) -> str:
    """Strip query params from URL to avoid leaking sensitive data."""
    return url.split("?")[0] if "?" in url else url


def instrument_requests(add_breadcrumb: Callable[..., None]) -> None:
    """
    Patch ``requests.Session.send`` to add automatic HTTP breadcrumbs.

    Safe to call even if ``requests`` is not installed (no-op).
    """
    try:
        import requests  # type: ignore[import-untyped]
    except ImportError:
        return

    original_send = requests.Session.send

    @functools.wraps(original_send)
    def patched_send(self: Any, request: Any, **kwargs: Any) -> Any:
        import time

        method = request.method or "GET"
        url = _sanitize_url(request.url or "")
        start = time.monotonic()
        try:
            response = original_send(self, request, **kwargs)
            duration_ms = int((time.monotonic() - start) * 1000)
            add_breadcrumb(
                type="http",
                message=f"{method} {url} -> {response.status_code}",
                level="error" if response.status_code >= 400 else "info",
                data={
                    "method": method,
                    "url": url,
                    "statusCode": response.status_code,
                    "durationMs": duration_ms,
                },
            )
            return response
        except Exception as e:
            duration_ms = int((time.monotonic() - start) * 1000)
            add_breadcrumb(
                type="http",
                message=f"{method} {url} -> failed: {type(e).__name__}",
                level="error",
                data={
                    "method": method,
                    "url": url,
                    "error": str(e),
                    "durationMs": duration_ms,
                },
            )
            raise

    requests.Session.send = patched_send  # type: ignore[assignment]


def instrument_logging(add_breadcrumb: Callable[..., None]) -> None:
    """
    Add a logging handler that creates breadcrumbs for WARNING+ log messages.

    Avoids infinite recursion by ignoring messages from the ``allstak`` logger
    namespace itself.
    """

    class BreadcrumbHandler(logging.Handler):
        def emit(self, record: logging.LogRecord) -> None:
            # Skip our own SDK logs to prevent infinite recursion
            if record.name.startswith("allstak"):
                return
            if record.levelno >= logging.WARNING:
                level = "error" if record.levelno >= logging.ERROR else "warn"
                try:
                    add_breadcrumb(
                        type="log",
                        message=record.getMessage(),
                        level=level,
                        data={"logger": record.name, "module": record.module},
                    )
                except Exception:
                    pass

    handler = BreadcrumbHandler()
    handler.setLevel(logging.WARNING)
    logging.getLogger().addHandler(handler)

File: packages/allstak-js/src/modules/auto-breadcrumbs.ts
/**
 * Automatic breadcrumb instrumentation for fetch and console in browser environments.
 *
 * These patches are safe: they only wrap if the globals exist and always
 * delegate to the original implementation.
 */

type AddBreadcrumbFn = (
  type: string,
  msg: string,
  level?: string,
  data?: Record<string, unknown>,
) => void;

/**
 * Wrap `globalThis.fetch` to record HTTP breadcrumbs for every request.
 */
export function instrumentFetch(addBreadcrumb: AddBreadcrumbFn): void {
  if (typeof globalThis.fetch !== 'function') return;

  const originalFetch = globalThis.fetch;

  globalThis.fetch = async function (
    input: RequestInfo | URL,
    init?: RequestInit,
  ): Promise<Response> {
    const method = init?.method?.toUpperCase() || 'GET';
    const url =
      typeof input === 'string'
        ? input
        : input instanceof URL
          ? input.href
          : input.url;
    // Strip query string to avoid leaking sensitive params
    const safePath = url.split('?')[0];

    const start = Date.now();
    try {
      const response = await originalFetch.call(this, input, init);
      const durationMs = Date.now() - start;
      addBreadcrumb(
        'http',
        `${method} ${safePath} -> ${response.status}`,
        response.status >= 400 ? 'error' : 'info',
        { method, url: safePath, statusCode: response.status, durationMs },
      );
      return response;
    } catch (err) {
      const durationMs = Date.now() - start;
      addBreadcrumb('http', `${method} ${safePath} -> failed`, 'error', {
        method,
        url: safePath,
        error: String(err),
        durationMs,
      });
      throw err;
    }
  };
}

/**
 * Wrap `console.warn` and `console.error` to record log breadcrumbs.
 */
export function instrumentConsole(addBreadcrumb: AddBreadcrumbFn): void {
  if (typeof console === 'undefined') return;

  const origWarn = console.warn;
  const origError = console.error;

  console.warn = function (...args: unknown[]) {
    addBreadcrumb('log', args.map(String).join(' '), 'warn');
    origWarn.apply(console, args);
  };

  console.error = function (...args: unknown[]) {
    addBreadcrumb('log', args.map(String).join(' '), 'error');
    origError.apply(console, args);
  };
}
